pkg/cloudprovider/aws/packing/packing.go

```diff
@@ -77,33 +90,76 @@ func (p *podPacker) Pack(ctx context.Context, constraints *cloudprovider.Constra
 	return packings, nil
 }
 
-// TODO filter instance types based on node contraints like availability zones etc.
-func (p *podPacker) getNodeCapacities(constraints *cloudprovider.Constraints) []*nodeCapacity {
+func (p *podPacker) getNodeCapacities(constraints *cloudprovider.Constraints) ([]*nodeCapacity, error) {
 	result := make([]*nodeCapacity, 0)
-	for _, nc := range nodeCapacities {
-		ncc := nc.Copy()
-		kubeletOverhead := binpacking.CalculateKubeletOverhead(ncc.total)
-		if ok := ncc.reserve(resources.Merge(constraints.Overhead, kubeletOverhead)); !ok {
-			zap.S().Errorf("Failed to reserve kubelet overhead for node capacity type %v", ncc.instanceType)
+
+	describeInstanceTypesInput := &ec2.DescribeInstanceTypesInput{
+		Filters: constraintsToDescribeInstanceTypesFilters(constraints),
+	}
+
+	err := p.ec2.DescribeInstanceTypesPagesWithContext(context.TODO(), describeInstanceTypesInput, func(page *ec2.DescribeInstanceTypesOutput, lastPage bool) bool {
+		for _, instanceTypeInfo := range page.InstanceTypes {
+			nc := instanceTypeInfoToNodeCapacity(*instanceTypeInfo)
+			kubeletOverhead := binpacking.CalculateKubeletOverhead(nc.total)
+			if ok := nc.reserve(resources.Merge(constraints.Overhead, kubeletOverhead)); !ok {
+				zap.S().Errorf("Failed to reserve kubelet overhead for node capacity type %v", nc.instanceType)
+			}
+			result = append(result, nc)
 		}
-		result = append(result, ncc)
+		return lastPage
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("fetching instance types using ec2.DescribeInstanceTypes, %w", err)
 	}
-	return result
+	return result, nil
 }
 
-type packingResult struct {
-	packed   []*v1.Pod
-	unpacked []*v1.Pod
+func constraintsToDescribeInstanceTypesFilters(constraints *cloudprovider.Constraints) []*ec2.Filter {
```

Comment:
re: naming, I'm a fan of this style
```
filterFrom(constarints *cloudprovider.Constraints)
```
inline it will read like
```
filterFrom(constraints)
```

Reply:
I like that naming style, but the problem here is that the filters are specifically for the DescribeInstanceTypes call. So the return type doesn't give you enough information to infer where you can use it. 

Would this be ok? 

```
describeInstanceTypesFiltersFrom(constraints)
```

Reply:
SGTM. I think once we move this into its own file (i.e. instance_type_provider), that sort of information may become implied by context. In this context, I agree.